HIP/hip-435.md

```diff
@@ -0,0 +1,247 @@
+  HashObject endObjectRunningHash = 4; 
+
+  // The block number associated with this period.
+  int64 blockNumber = 5; 
+
+  // List of the hashes of all the Sidecar record files created for the same period.
+  // Allows multiple sidecar files to be linked to this RecordStreamFile
+  repeated SidecarMetadata sidecars = 6; 
+}
+```
+
+```
+enum SidecarType {
+  ContractStateChange = 1;
+  ContractAction = 2;
+  // Other future categories
+}
+```
+
+```
+message SidecarMetadata {
+  bytes hash = 1;
+  long id = 2;
+  repeated SidecarType types = 3;
+}
+```
+
+```
+message HashObject {
+  enum HashAlgorithm {
+    SHA-384 = 0
+  }
+  HashAlgorithm algorithm = 0;
+  int length = 1;
+  bytes hash = 2;
+}
+```
+
+`RecordStreamFile` is a newly defined protobuf message which will be used to serialise all `RecordFileObject`s that are part of the same period into record stream files. Periods are intervals of `hedera.recordStream.logPeriod` seconds used to group `RecordFileObject`s into record stream files.
+
+The record signature file that is created for each record stream file signs the hash of the entire corresponding stream file. The list of sidecar file hashes is included in the record stream file. This way mirror nodes or any interested party can download the record stream file and all sidecar files and verify that:
+1. repeated bytes SidecarFileHash is correct; 
+2. the signature file signed the correct hash of the entire record stream file;
+
+```
+message TransactionSidecarRecord {
+  // Consensus timestamp will be the same as the consensus timestamp of the 
+  // transaction the side car is related to. This offers a convenient 
+  // way to match record to sidecar. 
+  Timestamp consensusTimestamp = 1;
+  oneof {
+    repeated ContractStateChange = 2; // Example sidecar type
+    repeated ContractAction = 3; // Example sidecar type
+    // Other future categories
+  }
+}
+```
+
+`TransactionSidecarRecord` is a protobuf message which will be used to create sidecar records complementing `TransactionRecord` and storing additional information about a transaction execution. 
+
+```
+message SidecarFile {
+  repeated TransactionSidecarRecord sidecarRecords = 2;
+}
+```
+
+The `TransactionSidecarRecord`s created by Hedera nodes will be externalised in `SidecarFile` files grouped by the type of the records. Files will be created using the SidecarFile protobuf holding all the `TransactionSidecarRecords` of a same type (e.g. `ContractStateChange`, `ContractAction`) related to `RecordFileObject`s created for the same period.
+
+The `SidecarFile`s will be uploaded into a separate folder in the cloud bucket. Doing so will reduce the number of rows Mirror nodes have to search in S3, reduce S3 list response size, and avoid having to filter them out locally. Probably `recordstreams/record0.0.3/sidecar`.
+
+```
+message SignatureFile {
+  
+  // version of HAPI that was used to serialise this file.
+  int fileVersion = 1;
```

Comment:
This wouldn't be related to HAPI version since that tracks services tags. Record file versions change at an infrequent rate. This would be set to 6. 

```suggestion
  // The signature file version that was used to serialise this file.
  int version = 1;
```